rfcs/2023-radix-form-primitive.md

```diff
@@ -0,0 +1,301 @@
+
+Similary to `Form.ClientMessage`, all accessibility relating to server errors is handled by the primitive:
+
+- when one or more server-side error message display, they are automatically associated to their matching control and focus is moved to the first invalid control.
+- when there are only global server errors, focus is moved to the submit button and the message is associated with it so screen readers announce it.
+
+Let's see the same example as above but with added server-side error handling:
+
+```jsx
+import * as React from 'react';
+import * as Form from '@radix-ui/react-form';
+
+function Page() {
+  const [serverErrors, setServerErrors] = React.useState({});
+
+  return (
+    <Form.Root
+      onSubmit={(event) => {
+        // `onSubmit` only triggered if it passes client-side validation
+        event.preventDefault();
+        const data = Object.fromEntries(new FormData(event.currentTarget));
+
+        // maybe do an async server call and gather some errors from the server to display
+        submitForm(data)
+        .then(() => …)
+        /**
+         * Map errors from your server response into the expected format of `serverErrors` (see type below)
+         * For example:
+         * {
+         *   email: [ { code: 'email_invalid', 'This is not a valid email.' } ],
+         *   global: [ { code: 'server_error', 'Something went wrong.' }]
+         * }
+         */
+        .catch((errors) => setServerErrors(mapServerErrors(errors)));
+      }}
+      serverErrors={serverErrors}
+      onServerErrorsChange={setServerErrors}
+    >
+      <Form.Field name="name">
+        <Form.Label>Full name</Form.Label>
+        <Form.Control />
+        <Form.ClientMessage type="missingValue">Please enter your name.</Form.ClientMessage>
+        <Form.ServerMessage />
+      </Form.Field>
+
+      <Form.Field name="email">
+        <Form.Label>Email address</Form.Label>
+        <Form.Control type="email" />
+        <Form.ClientMessage type="missingValue">Please enter your email.</Form.ClientMessage>
+        <Form.ClientMessage type="typeMismatch">Please provide a valid email.</Form.ClientMessage>
+        <Form.ServerMessage /> {/* will yield "This is not a valid email." */}
+      </Form.Field>
+
+      <Form.Field name="age">
+        <Form.Label>Age</Form.Label>
+        <Form.Control type="number" min="18" max="57" step="1" />
+        <Form.ClientMessage type="missingValue">Please enter your age.</Form.ClientMessage>
+        <Form.ClientMessage type="rangeOverflow">You must be 18 or older.</Form.ClientMessage>
+        <Form.ClientMessage type="rangeUnderflow">You must be 57 or younger.</Form.ClientMessage>
+        <Form.ClientMessage type="stepMismatch">Please enter a whole number.</Form.ClientMessage>
+        <Form.ServerMessage />
+      </Form.Field>
+
+      <Form.Submit>Submit</Form.Submit>
+
+      <Form.ServerMessage /> {/* will yield "Something went wrong." */}
```

Comment:
I think I'd rather operate on codes returned from the server so that I can manage the translations in one place.

```tsx
<Form.ServerMessage type="email_taken">
  {/* react-intl */}
  <FormattedMessage defaultMessage="Account with email already exists" ... />
</Form.ServerMessage>
```

Reply:
@benoitgrelard yeah i also have concerns about the `Form.ClientMessage` and `Form.ServerMessage` being two separate components. from a progressive enhancement mindset, it feels strange to separate them.

i think i would rather have `Form.ValidationMessage` or similar that renders client-side and server-side, with a `defaultValid` prop or similar for specifying uncontrolled default state from server:

```jsx
<Form.ValidationMessage type="missingValue" defaultValid={!serverData.errors.name}>Name is required</Form.ValidationMessage>
```

Reply:
Yeah the separation between `ClientMessage` and `ServerMessage` is the part I'm the last happy with hence why it was one of the open questions. Glad folks are suggesting ideas for this too. Will iterate on them.

Thanks! 🙏